apps/dash-study-browser/app.py

```diff
@@ -0,0 +1,326 @@
+                        html.Div(
+                            className="bg-white",
+                            children=[
+                                html.H5("Animal data plot"),
+                                dcc.Graph(id="plot"),
+                            ],
+                        )
+                    ],
+                ),
+                dcc.Store(id="stored-data", storage_type="memory"),
+            ],
+        ),
+    ]
+)
+
+# Callback to generate error message
+# Also sets the data to be used
+# If there is an error use default data else use uploaded data
+@app.callback(
+    [
+        Output("error-message", "children"),
+        Output("study-dropdown", "options"),
+        Output("study-dropdown", "value"),
+    ],
+    [Input("upload-data", "contents")],
+)
+def update_error(contents):
+
+    error_message = None
+    study_data = default_study_data
+
+    # Check if there is uploaded content
+    if contents:
+        content_type, content_string = contents.split(",")
+        decoded = base64.b64decode(content_string)
+
+        # Try reading uploaded file
+        try:
+            study_data = pd.read_csv(io.StringIO(decoded.decode("utf-8")))
+
+            # Missing columns
+            missing_columns = {
+                "group_id",
+                "group_type",
+                "reading_value",
+                "study_id",
+            }.difference(study_data.columns)
+
+            if missing_columns:
+                error_message = html.Div(
+                    className="alert",
+                    children=["Missing columns: " + str(missing_columns)],
+                )
+                study_data = default_study_data
+
+        # Data is invalid
+        except Exception as error:
+            error_message = html.Div(
+                className="alert",
+                children=["That doesn't seem to be a valid csv file!"],
+            )
+            study_data = default_study_data
+
+    # Update Dropdown
+    options = []
+    if "test_article" in study_data.columns:
+        test_articles = study_data.test_article.unique()
+        for test_article in test_articles:
+            for study in study_data.study_id[
+                study_data.test_article == test_article
+            ].unique():
+                options.append(
+                    {"label": f"{test_article} (study: {study})", "value": study}
+                )
+    else:
+        for study in study_data.study_id.unique():
+            options.append({"label": study, "value": study})
+
+    options.sort(key=lambda item: item["label"])
+    value = options[0]["value"]
+
+    return error_message, options, value
+
+
+# Callback to generate study data
+@app.callback(
+    Output("plot", "figure"),
+    [Input("chart-type", "value"), Input("study-dropdown", "value")],
+    [State("upload-data", "contents"), State("error-message", "children")],
+)
+def update_output(chart_type, study, contents, error):
+    if error or not contents:
```

Comment:
Can you try putting the state of the error in a `dcc.Store` instead? This is kind of a weird pattern to use (checking to see if an HTML component's children is `None`).